desc/integrals/_bounce_utils.py

```diff
@@ -691,79 +679,93 @@ def fourier_chebyshev(theta, iota, alpha, num_transit):
+        enumerates field line ``alpha[i]``. Each Chebyshev series approximates
+        ``f`` over one toroidal transit.
 
     """
-    # When f = |B|, it is expected that Y > T.Y so the code immediately below
-    # is then up-sampling the Chebyshev resolution, which is good since the
-    # spectrum of |B| is wider than θ.
-
-    # θ at Chebyshev points, reshaped to (..., num transit * num points)
-    theta = T.evaluate(Y).reshape(*T.cheb.shape[:-2], T.X * Y)
-    zeta = jnp.broadcast_to(cheb_pts(Y, domain=T.domain), (T.X, Y)).ravel()
-
-    # f at Chebyshev points
-    f = _irfft2_non_uniform(
-        theta, zeta, f, n0, n1, domain1=(0, 2 * jnp.pi / NFP)
-    ).reshape(*T.cheb.shape[:-1], Y)
-    f = PiecewiseChebyshevSeries(cheb_from_dct(dct(f, type=2, axis=-1)) / Y, T.domain)
+    # Let m, n denote the poloidal and toroidal Fourier resolution. We need to
+    # compute a set of 2D Fourier series each on non-uniform tensor product grids
+    # of size |𝛉|×|𝛇| where |𝛉| = num alpha × num transit and |𝛇| = Y.
+    # Partial summation is more efficient than direct evaluation when
+    # mn|𝛉||𝛇| > mn|𝛇| + m|𝛉||𝛇| or equivalently n|𝛉| > n + |𝛉|.
+
+    zeta = cheb_pts(Y, domain=T.domain)
+    # Shape broadcasts with (num rho, Y, m_modes.size).
+    f = ifft_non_uniform(
+        zeta[:, jnp.newaxis],
+        f,
+        _modes=n_modes,
+        domain=(0, 2 * jnp.pi / NFP),
+        axis=-2,
+    )
+    # f at Chebyshev points in ζ on field lines
+    f = irfft_non_uniform(T.evaluate(Y), f[..., jnp.newaxis, :, :], m, _modes=m_modes)
+    f = PiecewiseChebyshevSeries(cheb_from_dct(dct(f, type=2, axis=-1) / Y), T.domain)
     return f
 
 
-def cubic_spline(T, f, Y, n0, n1, NFP=1, check=False):
-    """Cubic spline along field lines.
+def cubic_spline(T, f, Y, m, m_modes, n_modes, NFP=1, check=False):
+    """Compute cubic spline of ``f`` on field lines using fast transforms.
 
     Parameters
     ----------
     T : PiecewiseChebyshevSeries
-        Set of 1D Chebyshev spectral coefficients of θ along field line.
-        {θ_α : ζ ↦ θ(α, ζ) | α ∈ A} where A = (α₀, α₁, …, αₘ₋₁) is the same
-        field line as ``alpha``. Each Chebyshev series approximates θ over
-        one toroidal transit.
+        Set of 1D Chebyshev spectral coefficients of θ on field lines.
+        {θ_αᵢⱼ : ζ ↦ θ(αᵢⱼ, ζ) | αᵢⱼ ∈ Aᵢ} where Aᵢ = (αᵢ₀, αᵢ₁, ..., αᵢ₍ₘ₋₁₎)
+        enumerates field line ``alpha[i]``. Each Chebyshev series approximates
+        θ over one toroidal transit.
     f : jnp.ndarray
+        Shape broadcasts with (num rho, 1, n_modes.size, m_modes.size).
         Fourier transform of f(θ, ζ) as returned by ``Bounce2D.fourier``.
-        ``n0=grid.num_theta``, ``n1=grid.num_zeta``, ``NFP=grid.NFP``.
     Y : int
-        Number of knots per transit to interpolate ``f``.
-    n0 : int
+        Number of knots per toroidal transit to interpolate ``f``.
+        This number will be rounded down to an integer multiple of NFP.
```

Comment:
I confirmed this is the reason the baseline images for the neoclassical and fast ion confinement stuff changed. This can be confirmed easily by setting NFP to 1 everywhere in this function except for the line of code with `domain=(0, 2 * jnp.pi / NFP)`.